Feat: Make ID generator configurable

EdZou opened the pull request:
## Which problem is this PR solving?
The existing implementation of id generating only contains two functions while in the Java and Go repository, we provide an interface and default IdGenerator implementation.

## Short description of the changes
1. added id generator interface under opentelemetry-core repo
2. migrated original id generator functions and corresponding unit tests by implementing IdGenerator interface.
3. modified TracerConfig and Tracer to make it configurable for users.

anuraaga commented on packages/opentelemetry-core/src/platform/node/AWSXrayIdGenerator.ts:
```diff
@@ -0,0 +1,46 @@
+/*
```
In Java we have them in the main repo

https://github.com/open-telemetry/opentelemetry-java/tree/master/sdk_extensions/aws_v1_support

From what I understand currently this is being left up to the language sigs. So if JS says separate repo we'll do that :)

EdZou commented on packages/opentelemetry-api/src/platform/node/IdGenerator.ts:
```diff
@@ -14,23 +14,15 @@
- * Returns a random 16-byte trace ID formatted/encoded as a 32 lowercase hex
- * characters corresponding to 128 bits.
+ * IdGenerator provides an interface for generating Trace Id and Span Id
+ *
+ * The prototype IdGenerator generates random 16-byte trace ID and 8-byte span ID
```
make sense, will do it!

EdZou commented on packages/opentelemetry-api/src/platform/node/IdGenerator.ts:
```diff
@@ -14,23 +14,15 @@
- */
-export function randomSpanId(): string {
-  return crypto.randomBytes(SPAN_ID_BYTES).toString('hex');
+export interface IdGenerator {
```
Thank you for your comment and sorry about that, I am really not sure where should this interface be added. Because current id.ts is under platform folder, I simply followed it and added interface to api/.../platform folder. If there is any instruction for where it should be placed would be really helpful :)

EdZou replied:
Oh, you mentioned I should move interface outside of platform. Did you mean I should move it from ".../platform/node" to ".../platform" because both web and node version are using the same interface, or elsewhere? @dyladan

dyladan replied:
I would create a new place for it outside of `platform`

EdZou commented on packages/opentelemetry-api/src/platform/node/IdGenerator.ts:
```diff
@@ -14,23 +14,15 @@
-  return crypto.randomBytes(SPAN_ID_BYTES).toString('hex');
+export interface IdGenerator {
+  /** Generate Trace ID which is default to be 16 bytes */
+  GenerateTraceId(): string;
```
My fault, I will walk through it and modify the code, thank you!

EdZou commented on packages/opentelemetry-core/src/platform/node/AWSXrayIdGenerator.ts:
```diff
@@ -0,0 +1,46 @@
+const TRACE_ID_BYTES = 16;
+const TIME_BYTES = 4;
+
+export class AWSXrayIdGenerator implements api.IdGenerator {
```
Nice suggestion! I will add it

EdZou commented on packages/opentelemetry-core/test/platform/AWSXrayIdGenerator.test.ts:
```diff
@@ -0,0 +1,63 @@
+/*
```
Really sorry about the confusion, since I have no idea where to put vendor code, I temporarily put code and corresponding tests under the same repo as original id.ts and drafted a PR to get suggestions here.

EdZou commented on packages/opentelemetry-core/src/platform/node/index.ts:
```diff
@@ -17,7 +17,8 @@
 export * from './environment';
 export * from './hex-to-base64';
-export * from './id';
+export * from './RandomIdGenerator';
```
Since I am new to JS, it may be a dumb question: 
Do you mean when we write code under .../node repo and want to export it, we should also add an `export * from '../../opentelemetry-core/src/platform/node/RandonIdGenerator` in the "opentelemetry-web" index file to solve dependency problem and pass through CircleCI? 
Or you mean I should add code under "opentelemetry-core/src/platform/browser"?
I just read through the code under opentelemetry-web, and still not sure what to do.

dyladan replied:
When the project is compiled for web, the `node` folder is completely replaced by the `web` folder. This means if you do `import { MyClass } from './platform/node/myclass';`, in the web this will be transparently replaced with `import { MyClass } from './platform/web/myclass';`. So the exports from node and web need to have identical interfaces.

obecny replied:
You should not export anything from `platform/node` directly, all you have to do is to add a new stuff into corresponding `index.ts` in folder `platform/node/` and `platform/browser`. Platform itself is exported already so anything you add in those indexes will be exported.

EdZou replied:
Thank you for you instructions, I have already get this done!

EdZou commented on packages/opentelemetry-tracing/src/types.ts:
```diff
@@ -52,6 +53,9 @@ export interface TracerConfig {
   /** Resource associated with trace telemetry  */
   resource?: Resource;
+
+  /** Id Generator configuration */
```
Already changed and commit! Thank you for pointing out!

EdZou commented on packages/opentelemetry-core/test/platform/RandomIdGenerator.test.ts:
```diff
@@ -0,0 +1,67 @@
+const IdGenerator = new RandomIdGenerator();
+const IdGeneratorWithParameters = new RandomIdGenerator(8, 16);
+
+describe('IdGeneratorConstructor', () => {
```
I actually added beforeEach() keyword to separate obtaining value part and testing value part. Is that good or not?

EdZou commented:
> Few things.
> 
>     1. Move all interfaces to types.ts
> 
>     2. You should mimic the same behaviour for browser and for node
> 
>     3. When bundle is being compiled the compiler checks for "browser" field in package.json if you check it you will see that file `./src/platform/index.ts` is replaced with file `./src/platform/browser/index.ts`
> 
>     4. When you write tests you should write a test that is not platform dependent, it means you add this in folder `test`. When the tests will be run in node it will pickup the `id` from node and run tests using this implementation. When tests will be run in karma it will use the `browser` from package.json replace the appropriate files and will run the test in browser environment. It means you write one test that should pass in both environments.
> 
>     5. You are lacking here the whole implementation for browser
> 
> 
> Besides all I have open questions
> 
>     1. Is this required by spec ?
> 
>     2. What are you trying to achieve ? Is this to be able to inject your own id generators ?
> 
> 
> Concerns:
> 
>     1. The AWSXrayIdGenerator should not be a part of api, this can be either in core package (it might make sense) or in contrib repo

Thank you for your comment! I have already added web version of IdGenerator
For your question:
1. I am not sure whether it is required by spec.
2. Since we are going to implement AWS Xray IdGenerator, after discussion with my mentor, Anuraag, we believe provide an interface for not only AWS Xray IdGenerator will be good. Then I first add a interface and wrap original id generator function in a class.
Concerns:
Yes, I have already removed AWS part from this repo.
Again, thank you for your help!

dyladan commented on packages/opentelemetry-core/src/platform/browser/RandomIdGenerator.ts:
```diff
@@ -0,0 +1,67 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import {IdGenerator} from '../../trace/IdGenerator';
```
nit: Prefer spaces like this `{ IdGenerator }`

EdZou replied:
Thank you! I am working on it and already passed through the `npm run lint` test locally. Will make another commit soon

dyladan commented on packages/opentelemetry-core/src/platform/browser/RandomIdGenerator.ts:
```diff
@@ -0,0 +1,67 @@
+ */
+import {IdGenerator} from '../../trace/IdGenerator';
+
+declare type WindowWithMsCrypto = Window & {
```
Is `declare` needed? Was it there in the old `id.ts`?

EdZou replied:
Yes, I preserved most of the original web version code

dyladan commented on packages/opentelemetry-core/src/platform/browser/RandomIdGenerator.ts:
```diff
@@ -0,0 +1,67 @@
+  
+    return String.fromCharCode.apply(null, chars);
+  }
+}
```
newline at EOF

dyladan commented on packages/opentelemetry-core/src/platform/node/RandomIdGenerator.ts:
```diff
@@ -15,22 +15,26 @@
  */
 
 import * as crypto from 'crypto';
+import {IdGenerator} from '../../trace/IdGenerator';
```
same spacing nit

dyladan commented on packages/opentelemetry-tracing/src/Tracer.ts:
```diff
@@ -42,6 +42,7 @@ export class Tracer implements api.Tracer {
   readonly resource: Resource;
   readonly instrumentationLibrary: InstrumentationLibrary;
   readonly logger: api.Logger;
+  readonly idGenerator: IdGenerator;
```
make private, prefix with `_`

EdZou replied:
Nice suggestion, let me do it!

dyladan commented on packages/opentelemetry-tracing/src/types.ts:
```diff
@@ -52,6 +52,12 @@ export interface TracerConfig {
+
+  /**
+   * Generator of trace and span IDs 
+   * The prototype IdGenerator generates random 16-byte trace ID and 8-byte span ID
```
add default to tsdoc
```suggestion
   * The default idGenerator generates random ids
```

EdZou replied:
Actually I initially add this to tsdoc but my mentor Anuraag suggested me to add default info here

anuraaga replied:
Not quite sure why I'm summoned here :) But it seems to be a good suggestion for cleaning up the doc, can you apply it?

legendecas replied:
Is this conversation going to be addressed?

The PR is LGTM if the conversation has been resolved.

👍 Nice work!

dyladan replied:
It's a minor nit so it's not blocking for me, but I would prefer to see the change made :)

dyladan commented:
This looks a lot better after the changes, thanks

EdZou commented:
Really thank you for reviewing my code and answering my questions! @dyladan @obecny @anuraaga 
I believe current PR is ready for reiweing and be merged

dyladan commented on packages/opentelemetry-core/src/platform/browser/RandomIdGenerator.ts:
```diff
@@ -13,19 +13,17 @@
-import {IdGenerator} from '../../trace/IdGenerator';
+import { IdGenerator } from '../../trace/IdGenerator';
 
 declare type WindowWithMsCrypto = Window & {
```
Think this should work without declare. Can you try it?

EdZou replied:
Of course, already committed!

EdZou commented:
It seems we still need 2 more maintainer's review to get this PR merged, what should we do now? @dyladan

dyladan commented:
@EdZou you don't need maintainers, any approvers will do. @open-telemetry/javascript-approvers please take a look at this

EdZou commented:
@mwear @vmarchaud @OlivierAlbertini @markwolff @mayurkale22 @legendecas @naseemkullah Would you mind reviewing this PR to get it merged? Since Daniel and Obecny has approved it, we may still need 2 approvals to go.

dyladan commented:
@EdZou I already pinged that group 3 hours ago. Please be patient as many of those people are contributing in their own free time, or have limited time their work allows them to dedicate to open source efforts.

EdZou commented:
OMG, really sorry, I wrongly comprehend to let me take a look at the member of JS approver member lists. Really sorry for that.
